chain/actors/actor_storagemarket.go

```diff
@@ -0,0 +1,604 @@
+	nroot, err := vmctx.Storage().Put(&self)
+	if err != nil {
+		return nil, aerrors.HandleExternalError(err, "storing state failed")
+	}
+
+	aerr = vmctx.Storage().Commit(old, nroot)
+	if aerr != nil {
+		return nil, aerr
+	}
+
+	var outBuf bytes.Buffer
+	if err := out.MarshalCBOR(&outBuf); err != nil {
+		return nil, aerrors.HandleExternalError(err, "serialising output")
+	}
+
+	return outBuf.Bytes(), nil
+}
+
+func (st *StorageMarketState) validateDeal(vmctx types.VMContext, deal StorageDeal, providerWorker address.Address) aerrors.ActorError {
+	if vmctx.BlockHeight() > deal.Proposal.ProposalExpiration {
+		return aerrors.New(1, "deal proposal already expired")
+	}
+
+	if err := deal.Proposal.Verify(); err != nil {
+		return aerrors.Absorb(err, 2, "verifying proposer signature")
+	}
+
+	err := deal.Verify(providerWorker)
+	if err != nil {
+		return aerrors.Absorb(err, 2, "verifying provider signature")
+	}
+
+	// TODO: maybe this is actually fine
+	if vmctx.Message().From != providerWorker && vmctx.Message().From != deal.Proposal.Client {
+		return aerrors.New(4, "message not sent by deal participant")
+	}
+
+	// TODO: do some caching (changes gas so needs to be in spec too)
+	b, bnd, aerr := GetMarketBalances(vmctx.Context(), vmctx.Ipld(), st.Balances, deal.Proposal.Client, providerWorker)
+	if aerr != nil {
+		return aerrors.Wrap(aerr, "getting client, and provider balances")
+	}
+	clientBalance := b[0]
+	providerBalance := b[1]
+
+	if clientBalance.Available.LessThan(deal.Proposal.StoragePrice) {
+		return aerrors.Newf(5, "client doesn't have enough available funds to cover StoragePrice; %d < %d", clientBalance.Available, deal.Proposal.StoragePrice)
+	}
+
+	clientBalance = lockFunds(clientBalance, deal.Proposal.StoragePrice)
+
+	// TODO: REVIEW: Not clear who pays for this
+	if providerBalance.Available.LessThan(deal.Proposal.StorageCollateral) {
+		return aerrors.Newf(6, "provider doesn't have enough available funds to cover StorageCollateral; %d < %d", providerBalance.Available, deal.Proposal.StorageCollateral)
+	}
+
+	providerBalance = lockFunds(providerBalance, deal.Proposal.StorageCollateral)
+
+	// TODO: piece checks (e.g. size > sectorSize)?
+
+	bcid, aerr := setMarketBalances(vmctx, bnd, map[address.Address]StorageParticipantBalance{
+		deal.Proposal.Client: clientBalance,
+		providerWorker:       providerBalance,
+	})
+	if aerr != nil {
+		return aerr
+	}
+
+	st.Balances = bcid
+
+	return nil
+}
+
+type ActivateStorageDealsParams struct {
+	Deals []uint64
+}
+
+func (sma StorageMarketActor) ActivateStorageDeals(act *types.Actor, vmctx types.VMContext, params *ActivateStorageDealsParams) ([]byte, ActorError) {
+	var self StorageMarketState
+	old := vmctx.Storage().GetHead()
+	if err := vmctx.Storage().Get(old, &self); err != nil {
+		return nil, err
+	}
+
+	deals, err := amt.LoadAMT(types.WrapStorage(vmctx.Storage()), self.Deals)
+	if err != nil {
+		// TODO: kind of annoying that this can be caused by gas, otherwise could be fatal
```

Comment:
Yes, we need to investigate amt and hamt code to see how to make error handling in them better. It will probably be using `fmt.Errorf("whatver %w", error)` everywhere.

Reply:
This comment was inaccurate - HandleExternalError makes the error fatal if isn't an actor error - so it works as expected.